[SPARK-6435] spark-shell --jars option does not add all jars to classpath

tsudukim opened the pull request:
Modified to accept double-quotated args properly in spark-shell.cmd.

tsudukim committed:
[SPARK-6435] spark-shell --jars option does not add all jars to classpath

Modified to accept multiple jars in spark-shell.cmd.

SparkQA commented:
[Test build #29297 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29297/consoleFull) for   PR 5227 at commit [`04f4291`](https://github.com/apache/spark/commit/04f42914a9c5b0337895fbe0ff3df376b4d0966b).
 * This patch merges cleanly.

SparkQA commented:
[Test build #29297 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29297/consoleFull) for   PR 5227 at commit [`04f4291`](https://github.com/apache/spark/commit/04f42914a9c5b0337895fbe0ff3df376b4d0966b).
 * This patch **passes all tests**.

 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29297/
Test PASSed.

srowen commented:
I tend to trust your judgment on modifying these Windows scripts, especially if you've verified that this works where it did not before. CC @nishkamravi2 and @vanzin since this will touch the scripts also being modified in https://github.com/apache/spark/pull/5085

vanzin commented:
I think the real problem here is that `,` is a delimiter in Windows batch scripts and thus needs to be escaped:
http://ss64.com/nt/syntax-esc.html

So I think the correct fix would be to change `CommandBuilderUtils.quoteForBatchScript` to escape commas also. That would also avoid the temp file... which can cause problems if two jobs are being launched at the same time.

vanzin commented:
Ah, and even though jenkins doesn't run unit tests on Windows, it would be nice to update `SparkLauncherSuite` to add a unit test for this, since that will exercise this code path (regardless of which fix).

srowen commented:
OK @vanzin sounds like you recommend we not make this change. @tsudukim not sure if you feel comfortable trying to make the change he suggests?

vanzin commented:
Hi @tsudukim , are you planning on trying my suggestion? Otherwise I can make the change and test it.

tsudukim commented:
@vanzin I'm not sure I have got your suggestion right, but as I wrote in JIRA, I think this is not the Java side problem. 
https://issues.apache.org/jira/browse/SPARK-6435
Exactly the `,` is the problem. When Java application receives the arguments, they are already splitted by comma. If we want to escape comma, we should do it in batch side.

tsudukim commented:
Java application means the launcher `launcher\src\main\java\org\apache\spark\launcher\Main.java`.

tsudukim committed:
Merge branch 'master' into feature/SPARK-6435-2

Conflicts:
	bin/spark-class2.cmd

tsudukim committed:
- escaped comman and semicolon in CommandBuilderUtils.java
- added random string to the temporary filename
- double-quotation followed by `cmd /c` did not worked properly
- no need to escape `=` by `^`
- if double-quoted string ended with `\` like classpath, the last `\` is parsed as the escape charactor and the closing `"` didn't work properly

SparkQA commented:
[Test build #29590 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29590/consoleFull) for   PR 5227 at commit [`0d4dc41`](https://github.com/apache/spark/commit/0d4dc417d59338f897b0b8dbae846ab83f30a94b).

tsudukim commented:
Ah @vanzin, I didn't understand your suggestion. `CommandBuilderUtils` needs modified to escape comma.
But I think we still need to modify `spark-class2.cmd` as well.

When we execute `bin\spark-shell.cmd --jars "C:\Path to\jar1.jar,C:\Path to\jar2.jar"`, 
we get following with original `spark-class2.cmd` even if `CommandBuilderUtils` is fixed to escape comma.
`... --jars "C:\Path to\jar1.jar C:\Path to\jar2.jar" ...`
but we get with my PR `spark-class2.cmd`
`... --jars "C:\Path to\jar1.jar,C:\Path to\jar2.jar" ...`

Two job at the same time is surely a problem, so added random string to the filename to reduce the probability this problem occurs.

I know it's better to solve this problem without using temporary file if possible, but I have no idea now. If you've got any suggestion, please let me know.
I'd really like to replace all *.cmd to powershell rather than write ugly codes like this....

SparkQA commented:
[Test build #29590 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29590/consoleFull) for   PR 5227 at commit [`0d4dc41`](https://github.com/apache/spark/commit/0d4dc417d59338f897b0b8dbae846ab83f30a94b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29590/
Test FAILed.

tsudukim commented:
oops, forgot to include fixed test code.

tsudukim committed:
fixed test code for escaping '='.

SparkQA commented:
[Test build #29599 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29599/consoleFull) for   PR 5227 at commit [`1fee420`](https://github.com/apache/spark/commit/1fee420c46b1a9af0a5a8da074a55cd9cb8fd2ae).

srowen commented on launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java:
```diff
@@ -260,15 +260,14 @@ static String quoteForBatchScript(String arg) {
       quoted.appendCodePoint(cp);
     }
+    if (arg.codePointAt(arg.length() - 1) == '\\') {
+        quoted.append("\\");
```
This is indented too far. Why would a backslash need escaping only in the final position?

srowen replied:
Oh, I assumed you would escape all backslashes if they are the escape character. Is that not necessary?

tsudukim replied:
Backslash is escape character only when followed by `"`. I assumed it's the only case when the string ends with backslash.

srowen replied:
OK, so it's not a general escape character, got it.

srowen commented on launcher/src/main/java/org/apache/spark/launcher/Main.java:
```diff
@@ -101,12 +101,9 @@ public static void main(String[] argsArray) throws Exception {
    */
   private static String prepareWindowsCommand(List<String> cmd, Map<String, String> childEnv) {
-    StringBuilder cmdline = new StringBuilder("cmd /c \"");
+    StringBuilder cmdline = new StringBuilder("");
```
I assume you're sure about this, but we definitely don't always need `cmd /c` to run a windows command directly from Java? In Linux you have to have `bash` to interpret most of what we're used to typing on the command line.

tsudukim replied:
Why we are using `cmd.exe /c` in existing *.cmd is that the environment would be polluted by the executing commands.
But we already have *2.cmd and *.cmd calls *2.cmd by `cmd.exe /c`, so I think we don't need another `cmd.exe /c` when execute java.exe.
Rather, if we use `cmd.exe /c`, we face another problem.
We should execute like this command:
  `"C:\Program Files\Java\jdk1.7.0_67\bin\java.exe" -cp "C:\path\to\somewhere\;..." org.apache.spark.deploy.SparkSubmit ...`
But the Launcher returns this string:
  `cmd.exe /c ""C:\Program Files\Java\jdk1.7.0_67\bin\java.exe" -cp "C:\path\to\somewhere\;..." org.apache.spark.deploy.SparkSubmit ..."`
`cmd.exe /c` needs double-quoted command string, but the java.exe path and other arguments might be already double-quoted, so it the returned string is weird.

srowen replied:
OK I trust your judgment on this. This is confined to Windows support and I believe you're closest to this code. Note that you don't need the argument `""` now and there's still an indent issue above. Take a moment to see if it would be easy to add a test in `SparkLauncherSuite`; not sure if it is. Otherwise LGTM.

vanzin commented on launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java:
```diff
@@ -260,15 +260,14 @@ static String quoteForBatchScript(String arg) {
         quoted.append('"');
         break;
 
-      case '=':
```
Did you run SparkLauncherSuite on Windows? I added this specifically to fix an issue with that path (see 92a9cfb).

tsudukim replied:
I've run `SparkLauncherSuite` on Windows and it's OK.
If double-quotation is parsed properly, `=` in double-quotation is not need to be escaped.

vanzin commented:
I tried to figure out a solution that didn't involve a temp file, but failed. Guess I'll just delegate - change looks fine except for the one comment I left about testing.

srowen commented:
@tsudukim back to you regarding the test case. I'd like to get this in one way or the other since it's an important fix for Windows.

srowen commented:
@tsudukim Ping? I think this still has an outstanding question from Marcelo above that should be addressed.

tsudukim commented:
Ah, sorry to be late... I didn't have time to struggle with it. I'll do it this week.

srowen commented:
Is this one still active? I'd like to figure out whether this is stalled or might proceed.

tsudukim committed:
Merge branch 'master' of https://github.com/apache/spark into feature/SPARK-6435-2

Conflicts:
	bin/spark-class2.cmd

tsudukim commented:
I was checking about the `SparkLauncherSuite` on Windows as vanzin's comment, and faced some trouble. It seems not to related with this PR, but I'm not sure yet. Please give me more time a little. When I resolve the problem, I'll rebase this PR.

tsudukim committed:
removed unnecessary argument.

tsudukim commented:
The problem I mentioned was that the spark-shell.cmd which is called by `SparkLauncherSuite` somehow failed to launch test application.
It turned out to be caused by the limitation of Windows batch that the one command line must be shorter than 8192 characters. (The fullpath for classpath was long because I worked at a deep folder.)

So I assume all issue is now cleared up. Sorry for my late response.

SparkQA commented:
[Test build #719 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/719/consoleFull) for   PR 5227 at commit [`ac55787`](https://github.com/apache/spark/commit/ac55787f65a3cb4283a3d0a6671bb93dcf3abeda).

SparkQA commented:
[Test build #719 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/719/consoleFull) for   PR 5227 at commit [`ac55787`](https://github.com/apache/spark/commit/ac55787f65a3cb4283a3d0a6671bb93dcf3abeda).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

vanzin commented:
bq.  limitation of Windows batch that the one command line must be shorter than 8192 characters

This may be fixable is it's something we really care about, but not related to this change.

Given that the test passes, LGTM (based on my previous comments - didn't look at the changes again).

SparkQA commented:
[Test build #31039 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31039/consoleFull) for   PR 5227 at commit [`ac55787`](https://github.com/apache/spark/commit/ac55787f65a3cb4283a3d0a6671bb93dcf3abeda).

liukun1016 commented:
The same issue happens to the Mac OS. If multiple jar paths were specified, only the first one would be visible. Don't know if this issue was created and resolved or not.
